Invalid Timeout structure errors when running v0.12-based providers using v0.11 cli

<!--
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

If your issue relates to a specific Terraform provider, please open it in the provider's own repository. The index of providers is at https://github.com/terraform-providers .
-->

### Terraform Version
<!---
Run `terraform -v` to show the version, and paste the result between the ``` marks below.

If you are not running the latest version of Terraform, please try upgrading because your issue may have already been fixed.
-->

```
Terraform v0.11.8
```

### Terraform Configuration Files
<!--
Paste the relevant parts of your Terraform configuration between the ``` marks below.

For large Terraform configs, please use a service like Dropbox and share a link to the ZIP file. For security, you can also encrypt the files using our GPG public key.
-->

```hcl
resource "oci_email_sender" "test_sender" {
  #Required
  compartment_id = "${var.compartment_ocid}"
  email_address  = "${var.sender_email_address}"

  timeouts {
    create = "10m"
  }
}
```

### Debug Output
<!--
Full debug output can be obtained by running Terraform with the environment variable `TF_LOG=trace`. Please create a GitHub Gist containing the debug output. Please do _not_ paste the debug output in the issue, since debug output is long.

Debug output may contain sensitive information. Please review it before posting publicly, and if you are concerned feel free to encrypt the files using the HashiCorp security public key.
-->
...
oci_email_sender.test_sender - *terraform.NodePlannableResourceInstance
2019-01-30T10:54:04.888-0800 [DEBUG] plugin.terraform-provider-oci: 2019/01/30 10:54:04 [ERROR] Invalid timeout structure: []map[string]interface {}
2019/01/30 10:54:04 [DEBUG] Resource state not found for "oci_dns_record.record-txt": oci_dns_record.record-txt
2019/01/30 10:54:04 [DEBUG] ReferenceTransformer: "oci_dns_record.record-txt" references: []
2019/01/30 10:54:04 [DEBUG] Resource state not found for "data.oci_email_suppressions.test_suppressions": data.oci_email_suppressions.test_suppressions
2019/01/30 10:54:04 [DEBUG] Resource state not found for "data.oci_dns_records.records": data.oci_dns_records.records
2019/01/30 10:54:04 [TRACE] Graph after step *terraform.AttachStateTransformer:

data.oci_dns_records.records - *terraform.NodePlannableResourceInstance
2019/01/30 10:54:04 [TRACE] Graph after step *terraform.TargetsTransformer:

data.oci_dns_records.records - *terraform.NodePlannableResourceInstance
2019/01/30 10:54:04 [DEBUG] ReferenceTransformer: "data.oci_dns_records.records" references: []
2019/01/30 10:54:04 [DEBUG] Interpolating computed list element attribute records (74D93920-ED26-11E3-AC10-0800200C9A66)
2019/01/30 10:54:04 [DEBUG] plugin: waiting for all plugin processes to complete...

Error: Error running plan: 1 error(s) occurred:

* oci_email_sender.test_sender: 1 error(s) occurred:

* oci_email_sender.test_sender: [ERR] Error decoding timeout: Invalid Timeout structure found


### Expected Behavior
<!--
What should have happened?
-->
Running `terraform plan` should have succeeded. It succeeds when using a v0.11-based provider.

### Actual Behavior
<!--
What actually happened?
-->
Running `terraform plan` results in `oci_email_sender.test_sender: [ERR] Error decoding timeout: Invalid Timeout structure found` error.


### Steps to Reproduce
<!--
Please list the full steps required to reproduce the issue, for example:
1. `terraform init`
2. `terraform apply`
-->
terraform init
terraform plan

### Additional Context
<!--
Are there anything atypical about your situation that we should know? For example: is Terraform running in a wrapper script or in a CI system? Are you passing any unusual command line options or environment variables to opt-in to non-default behavior?
-->
This scenario involves v0.11 Terraform CLI that loads a provider built on top of `pluginsdk-v0.12-early2`

### References
<!--
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:

- #6017

-->

--- a/helper/schema/provider_test.go
+++ b/helper/schema/provider_test.go
@@ -278,7 +278,7 @@ func TestProviderValidate(t *testing.T) {
 	}
 }
 
-func TestProviderDiff_timeoutInvalidType(t *testing.T) {
+func TestProviderDiff_legacyTimeoutType(t *testing.T) {
 	p := &Provider{
 		ResourcesMap: map[string]*Resource{
 			"blah": &Resource{
@@ -298,6 +298,48 @@ func TestProviderDiff_timeoutInvalidType(t *testing.T) {
 	invalidCfg := map[string]interface{}{
 		"foo": 42,
 		"timeouts": []map[string]interface{}{
+			map[string]interface{}{
+				"create": "40m",
+			},
+		},
+	}
+	ic, err := config.NewRawConfig(invalidCfg)
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	_, err = p.Diff(
+		&terraform.InstanceInfo{
+			Type: "blah",
+		},
+		nil,
+		terraform.NewResourceConfig(ic),
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestProviderDiff_invalidTimeoutType(t *testing.T) {
+	p := &Provider{
+		ResourcesMap: map[string]*Resource{
+			"blah": &Resource{
+				Schema: map[string]*Schema{
+					"foo": {
+						Type:     TypeInt,
+						Optional: true,
+					},
+				},
+				Timeouts: &ResourceTimeout{
+					Create: DefaultTimeout(10 * time.Minute),
+				},
+			},
+		},
+	}
+
+	invalidCfg := map[string]interface{}{
+		"foo": 42,
+		"timeouts": []interface{}{
 			map[string]interface{}{
 				"create": "40m",
 			},

--- a/helper/schema/resource_timeout.go
+++ b/helper/schema/resource_timeout.go
@@ -63,7 +63,27 @@ func (t *ResourceTimeout) ConfigDecode(s *Resource, c *terraform.ResourceConfig)
 	}
 
 	if raw, ok := c.Config[TimeoutsConfigKey]; ok {
-		if timeoutValues, ok := raw.(map[string]interface{}); ok {
+		var rawTimeouts []map[string]interface{}
+		switch raw := raw.(type) {
+		case map[string]interface{}:
+			rawTimeouts = append(rawTimeouts, raw)
+		case []map[string]interface{}:
+			rawTimeouts = raw
+		case string:
+			if raw == config.UnknownVariableValue {
+				// Timeout is not defined in the config
+				// Defaults will be used instead
+				return nil
+			} else {
+				log.Printf("[ERROR] Invalid timeout value: %q", raw)
+				return fmt.Errorf("Invalid Timeout value found")
+			}
+		default:
+			log.Printf("[ERROR] Invalid timeout structure: %#v", raw)
+			return fmt.Errorf("Invalid Timeout structure found")
+		}
+
+		for _, timeoutValues := range rawTimeouts {
 			for timeKey, timeValue := range timeoutValues {
 				// validate that we're dealing with the normal CRUD actions
 				var found bool
@@ -108,14 +128,6 @@ func (t *ResourceTimeout) ConfigDecode(s *Resource, c *terraform.ResourceConfig)
 			}
 			return nil
 		}
-		if v, ok := raw.(string); ok && v == config.UnknownVariableValue {
-			// Timeout is not defined in the config
-			// Defaults will be used instead
-			return nil
-		}
-
-		log.Printf("[ERROR] Invalid timeout structure: %T", raw)
-		return fmt.Errorf("Invalid Timeout structure found")
 	}
 
 	return nil

--- a/helper/schema/resource_timeout_test.go
+++ b/helper/schema/resource_timeout_test.go
@@ -129,6 +129,45 @@ func TestResourceTimeout_ConfigDecode(t *testing.T) {
 	}
 }
 
+func TestResourceTimeout_legacyConfigDecode(t *testing.T) {
+	r := &Resource{
+		Timeouts: &ResourceTimeout{
+			Create: DefaultTimeout(10 * time.Minute),
+			Update: DefaultTimeout(5 * time.Minute),
+		},
+	}
+
+	raw, err := config.NewRawConfig(
+		map[string]interface{}{
+			"foo": "bar",
+			TimeoutsConfigKey: []map[string]interface{}{
+				{
+					"create": "2m",
+					"update": "1m",
+				},
+			},
+		})
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+	c := terraform.NewResourceConfig(raw)
+
+	timeout := &ResourceTimeout{}
+	err = timeout.ConfigDecode(r, c)
+	if err != nil {
+		t.Fatalf("Expected good timeout returned:, %s", err)
+	}
+
+	expected := &ResourceTimeout{
+		Create: DefaultTimeout(2 * time.Minute),
+		Update: DefaultTimeout(1 * time.Minute),
+	}
+
+	if !reflect.DeepEqual(timeout, expected) {
+		t.Fatalf("bad timeout decode.\nExpected:\n%#v\nGot:\n%#v\n", expected, timeout)
+	}
+}
+
 func TestResourceTimeout_DiffEncode_basic(t *testing.T) {
 	cases := []struct {
 		Timeout  *ResourceTimeout

